README.md
# Twitter Logging Service

## Background

Twitter created a framework for logging in order to fulfill the following requirements:

- fast (no blocking the main thread)
- thread safe
- as easy as `NSLog` in most situations
- support pluggable "output streams" to which messages will be delivered
- "output streams" filter messages rather than global filtering for all "output streams"
- able to categorize log messages (log channels)
- able to designate importance to log messages (log levels)
- require messages to opt-in for persisted logs (a security requirement, fulfilled by using the _context_ feature of *TLS*)

Twitter has been using *Twitter Logging Service* since January 2014 with minimal changes.  We've decided to share it with the developer community.

## List of good alternative logging frameworks

If *Twitter Logging Service* doesn't meet your needs, there are many great logging frameworks available, including the following high quality and well maintained projects:

- CocoaLumberjack
- SwiftyBeaver
- Apache Logging Services

## Architecture

There are 3 components to consider:
  1. the log message and its context
  2. the logging service instance or singleton
  3. the output stream(s)

The log message is sent to the logging service which provides the message to each output stream.

The logging service is configured by adding discrete output streams.  Output streams encapsulate their own behavior and decisions, including filtering and logging messages.  For instance, logging can mean printing to console with NSLog, writing to a file on disk, or sending the message to a remote server.

Message arguments don't need to be evaluated if the message is going to be filtered out.  This avoids expensive, synchronous execution of argument evaluation.  The message is then packaged with context before it is sent to the logging service.  Context includes information such as the log level, log channel, file name, function name, line number and timestamp.

The logging service marshals the message and its context to a background queue for processing by all available output streams.  Streams can then filter or output the message.

## Installation
### CocoaPods

[CocoaPods](http://cocoapods.org) is a dependency manager for Cocoa projects. You can install it using the following command:
```bash
    $ gem install cocoapods
```

To integrate TwitterLoggingService into your Xcode project using CocoaPods, specify it in your `Podfile`:
```ruby
    platform :ios, '8.0'
    use_frameworks!

    target "MyApp" do
        pod 'TwitterLoggingService', '~> 2.9.0'
    end
```

## Usage

`TLSLog.h` is the principal header for using *TwitterLoggingService*.  Just include `TLSLog.h` or `@import TwitterLoggingService`.

```c
    // The primary macros for *TwitterLoggingService*

    TLSLogError(channel, ...)          // Log at the TLSLogLevelError level
    TLSLogWarning(channel, ...)        // Log at the TLSLogLevelWarning level
    TLSLogInformation(channel, ...)    // Log at the TLSLogLevelInformation level
    TLSLogDebug(channel, ...)          // Log at the TLSLogLevelDebug level
```

For each macro in the `TLSLog` family of macros, `TLSCanLog` is called first to gate whether the actual
logging should occur.  This saves us from having to evaluate the arguments to the log message and can provide a win in performance when calling a `TLSLog` macro that will never end up being logged.  For more on `TLSCanLog` see `Gating TLSLog messages` below.

## TLSLog Core Macro

```c
    #define TLSLog(level, channel, ...)
```

`TLSLog` is the core macro and takes 3 parameters: a `TLSLogLevel` level, an `NSString` channel and then an `NSString` format with variable formatting arguments.
The level and channel parameters are used to filter the log message per `TLSOutputStream` in the `TLSLoggingService` singleton.  Providing `nil` as the channel argument to any logging macro, function or method will result in the message not being logged.

## Logging Channels, Levels and Context Objects

# Channels

The logging channel of a log message is an arbitrary string and acts as a tag to that message to further help identify what the message relates to.
Channels can help to quickly identify what a log message relates to in a large code base, as well as provide a mechanism for filtering.
A `TLSOutputStream` can filter based on the logging channel in its implementation of `tls_shouldFilterLevel:channel:contextObject:`.
Providing a `nil` channel to a log statement has the effect of not logging that message.

Examples of potential logging channels: @"Networking" for the networking stack, @"SignUp" for an app’s signup flow,  `TLSLogChannelDefault` as a catch all default logging channel, and @"Verbose" for anything you just want to log for the helluvit.

# Levels

The enum `TLSLogLevel` specifies 8 logging levels in accordance with the *syslog* specification for logging.
For practical use, however, only 4 log levels are used: `TLSLogLevelError`, `TLSLogLevelWarning`, `TLSLogLevelInformation` and `TLSLogLevelDebug`.
Each log message has a specified logging level which helps quickly identify its level, `TLSLogLevelEmergency` (or `TLSLogLevelError` in practice) is the most important while `TLSLogLevelDebug` is the least.
`TLSOutputStream` instances can filter a log message by its log level (in combination with its logging channel and context object) by implementing `tls_shouldFilterLevel:channel:contextObject:`.

An implementation detail to keep in mind w.r.t. logging levels is that `TLSLogLevelDebug` is ALWAYS filtered out in non-`DEBUG` builds.

# Context Objects

Though the `TLSLog` macros do not have a *context object* parameter, one can provide a *context object* to the `TLSLogging` APIs in order to provide additional context to custom `TLSOutputStream`s.
The *context object* will carry through the `TLSLoggingService` so that it is available to all `TLSOutputStream` instances.  The *context object* can be used to filter in the `tls_shouldFilterLevel:channel:contextObject:` method.  The *context object* can also be used for additional information in the logging of a message since it carries to the `TLSLogMessageInfo` object that's passed to `tls_outputLogInfo:`.

This *context object* provides near limitless extensibility to the `TLSLogging` framework beyond the basics of filtering and logging based on a logging level and logging channel.  Twitter uses the *context object* as a way to secure log messages from leaking to output streams that should not log messages unless explicitely told to do so, thus protecting Personally Identifiable Information from being logged as a default behavior.

## Setup

Setting up your project to use *TwitterLoggingService*:

1) Add the *TwitterLoggingService* XCode project as a subproject of your XCode project.

2) Add the *libTwitterLoggingService.a* library or *TwitterLoggingService.framework* framework as a dependency in your XCode project.

3) Set up your project to build the *TwitterLoggingService* project with `DEBUG=1` in debug builds and `RELEASE=1` in release builds.

4) Set up the `TLSLoggingService` singleton on application startup (often in `application:didFinishLaunchingWithOptions:` of your `UIApplication`'s delegate for iOS).

```
    @import TLSLoggingKit;

    // ...

    - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)options
    {
        // ...

        // Set up the Twitter Logging Service!
        TLSLoggingService *logger = [TLSLoggingService sharedInstance];
    #if DEBUG
        if ([TLSOSLogOutputStream supported]) {
            [logger addOutputStream:[[TLSOSLogOutputStream alloc] init]];
        } else {
            [logger addOutputStream:[[TLSNSLogOutputStream alloc] init]];
        }
    #endif
        [logger addOutputStream:[[TLSFileOutputStream alloc] initWithLogFileName:@"appname-info.log"]];

        // ...
    }

    // ...

    // Someplace else in your project
    - (void)foo
    {
        //  ...

        if (error) {
            TLSLogError(TLSLogChannelDefault, @"Encountered an error while performing foo: %@", error);
        } else {
            TLSLogInformation(@"Verbose", @"foo executed flawlessly!");
        }

        // ...
    }
```

## Best Practices

As a best practice follow these simple guidelines:

1) Any user sensitive information should not be logged to output streams that persist messages (including being sent over the network to be saved). You can configure your output stream to filter out logs to these sensitive channels.  Or do the inverse, and only permit certain "safe" channels to be logged.  Twitter has elected to use the pattern where only explicitely "safe" messages (designated via a custom context object) are logged to output streams that will persist. If in doubt, you can log to the `TLSLogLevelDebug` log level, which is only ever logged in `DEBUG` builds.

2) Configure `DEBUG` builds to have a console output like `TLSNSLogOutputStream` or `TLSStdErrOutputStream` - but add only 1 or you'll spam the debug console.

3) Configure `RELEASE` builds to not use the console output stream.

4) Add `Crashlytics` to your project and add a subclass of the `TLSCrashlyticsOutputStream` to `TLSLoggingService` instead of using `CLSLog`, `CLSNSLog` or `CLS_LOG`. You MUST subclass `TLSCrashlyticsOutputStream`.

## TLSLogChannelApplicationDefault function

```
    FOUNDATION_EXTERN NSString *TLSLogChannelApplicationDefault() __attribute__((const));
    #define TLSLogChannelDefault TLSLogChannelApplicationDefault()
```

Retrieve a channel based on the application.  You can use this as a default channel.

Loads and caches the channel name in the following order of descending priority:

1) `kCFBundleNameKey` of main bundle

2) `kCFBundleExecutableKey` of main bundle

3) Binary executable name

4) `@"Default"`

The default channel is available as a convenience for quick logging.  However, it is recommended to always have concrete, well-defined logging channels to which output is logged (e.g. "Networking", "UI", "Model", "Cache", et al).

## TLSLog Helper Functions

There are a number of **TLSLog Helper Functions** and they all accept as a first parameter a `TLSLoggingService`.
If `nil` is provided for the *service* parameter, the shared `[TLSLoggingService sharedInstance]` will be used.
All `TLSLog` macros use `nil` for the *service* parameter, but if there is different instance to be used, these helper functions support that.
As an example, Twitter extends *TwitterLoggingService* with its own set of macros so that a context is provided that defines the duration for which a message can be safely retained (e.g. to avoid retaining sensitive information), and uses custom macros that call these helper functions.

## Gating TLSLog messages

```
    BOOL TLSCanLog(TLSLoggingService *service, TLSLogLevel level, NSString *channel, id contextObject); // gate for logging TLSLog messages
```

At the moment, `TLSCanLog` evaluates two things (*contextObject* is currently ignored): the cached permitted log levels and the cached not permitted log channels.
A log message can log given the desired *level* is permitted by the internal cache of known permitted `TLSLogLevel`s based on the `outputStreams` of `TLSLoggingService`
AND the given log *channel* has not been cached as a known to be an *always off* channel (for `TLSLOGMODE=1` that is, see below for different behaviors).

## TLSCANLOGMODE build setting

*TwitterLoggingService* supports being compiled in one of 3 different modes:

* `TLSCANLOGMODE=0`
* `TLSCanLog` will always return `YES`
* Log arguments always evaluate, which can be inefficient for args that won't log
* `TLSCANLOGMODE=1`
* `TLSCanLog` will base its return value on cached insight into what can and cannot be logged
* This will save on argument evalution at the minimal cost of a quick lookup of cached information
* This is the default if `TLSCANLOGMODE` is not defined
* `TLSCANLOGMODE=2`
* `TLSCanLog` will base its return value on the filtering behavior of all the registered output streams
* This will save on argument evalution but requires an expensive examination of all output streams

# License

Copyright 2013-2020 Twitter, Inc.

Licensed under the Apache License, Version 2.0: https://www.apache.org/licenses/LICENSE-2.0

# Security Issues?

Please report sensitive security issues via Twitter's bug-bounty program (https://hackerone.com/twitter) rather than GitHub.


CHANGELOG.md
# Twitter Logging Service Change Log

## Info

**Document version:** 2.9.0

**Last updated:** 08/06/2020

**Author:** Nolan O'Brien

## History

### 2.9.0 (08/06/2020)

- Drop support for iOS 7, 8 & 9

### 2.8.5 (07/21/2020)

- Improved Swift interface

### 2.8.1 (07/01/2020)

- Convert private static C functions (that take `self` as an argument) to Objective-C `direct` methods
  - Same low overhead, better calling syntax

### 2.8.0 (06/08/2020)

- Update `initWithOutError:` to `initAndReturnError:` for `TLSRollingFileOutputStream`
  - More idiomatic
  
### 2.7.1 (04/12/2020)
 
 - Increase use of  `@autoreleasepool` to keep memory management tighter
 - Improve at launch perf by deferring the pruning of old files for `TLSRollingFileOutputStream`
 - Other miscellaneous bug fixes and cleanup

### 2.7.0 (06/28/2019)

- Add support for capturing `os_log` logs executed within a running app.
  - __TLSExt__ is provided as the interface and is not a part of __TLS__ proper.  This is because __TLSExt__ makes use of private Apple frameworks and would lead to app store rejection.
  - So, the consumer is responsible to compiling and linking the `TLSExt*.h|m` files themselves and must only do so with non-production builds.
  - This can be of immense help when dogfooding with developers and teammates with an Enterprise distribution (not Test Flight and not production).

### 2.6.0 (06/11/2019)

- Add options to composing a log message string from `TLSLogMessageInfo`
  - `TLSComposeLogMessageInfoOptions` provides options for what components to put in the output `composeFormattedMessageWithOptions:` string

### 2.5.0 (09/13/2016)

- Add options to _TLS_ with `TLSLogMessageOptions`
  - Provides support to explicitely ignore the maximum message length cap
  - Change `maximumSafeMessageLength` to truncate by default instead of discard when message length is too long and there is not delegate

### 2.4.0 (07/28/2016)

- Create `TLSLoggingServiceDelegate` protocol.  Currently delegates the decision of what to do when a log message with an unsafe length is encountered.

### 2.3.0 (06/30/2016)

- Add `os_log` support (iOS 10+ and macOS 10.12+) with `TLSOSLogOutputStream`

### 2.2.1 (05/31/2016)

- Add cap to log message sizes.  Will also fire a notification that can be observed to identify where the message that was too large was logged from.

### 2.2.0   (05/19/2016)

- Remove `TLSFileFunctionLine` struct since it is too easy to make mistakes such as constructing the struct on the stack with stack C-string values then accessing copies of the struct from other threads that should not have references to the stack C-string values. 

### 2.1.1   (05/04/2016)

- Add `threadName` to `TLSLogMessageInfo` for additional context

### 2.1.0   (03/23/2016)

- Refactor coding style/conventions to be better aligned with open source best practices
- Absolute minimal executable code changes

### 2.0.0   (02/25/2016)

- Rename `TFNLogging` to `TwitterLoggingService`

### 1.2.5   (02/03/2016)

- Add Swift support
- Simplify _Crashlytics_ support by delegating responsibility of calling `CLSLog` to the subclass of `TLSCrashlyticsOutputStream`

### 1.2.1   (09/11/2015)

- Optimize log message filtering by moving quick filter checks to a concurrent queue

### 1.2.0   (06/12/2014) - Kirk Beitz

- Made class `TLSFileOutputStream` more abstract as a base + protected implementation
- no longer implements @protocol `TLSDataRetrieval`
- keeps the generic readonly @property 'constants', but makes them @protected
- implements one default initializer taking a logging directory and a file name
- implements one convenience initializer taking a file name and making use of the default logging directory
- makes 'init' NS_UNAVAILABLE
- keeps the public `defaultLogFileDirectory` class method & the `tls_outputLogInfo:` and `tls_flush` methods
- keep the several protected `(void)write` methods
- refactor some portions of `(instancetype)initWithLogFileDirectoryPath:logFilePrefix:maxLogFiles:maxBytesPerLogFile:error:` to new protected `createLogFileDirectoryPath:error:` & `openLogFile:error:`

- Made new protocol `TLSFileOutputStreamEvent` based on methods that had been "abstract" and "overrideable" in the old TLSFileOutputStream
- makes use of the new typedef `TLSFileOutputEvent` for the first argument of all functions

- Made new class `TLSRollingFileOutputStream` as a concrete implementation of `TLSFileOutputStream`
- copied over all of the old initializers from `TLSFileOutputStream`
- copied over from `TLSFileOutputStream` the @property items that were specific to rolling file output stream
- interface creates new NS_ENUM `TLSRollingFileOutputEvent` (based upon `TLSFileOutputEvent` and override-able `fileStreamEvent` methods
- takes over implementation of protocol `TLSDataRetrieval`
- takes over implementation of `fileOutputEventBegan/Finished/Failed` methods via implementation of the @protocol `TLSFileOutputStreamEvent`

### 1.1.0   (04/09/2014)

- Remove `permittedLoggingLevels` and `shouldFilterChannelsThatAreOff` to completely decouple the filtering from the `TLSLoggingService`.  All output streams control their own destiny now.

### 1.0.1   (01/16/2014)

- Expand maximum bytes per log file from 128MB to 1GB.
- Add `flushAfterEveryWriteEnabled` property to `TLSLoggingService`.

### 1.0.0  (01/01/2014)

- Initial production release


Classes/TLSDeclarations.h
//
//  TLSDeclarations.h
//  TwitterLoggingService
//
//  Created on 12/11/13.
//  Copyright (c) 2016 Twitter, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//          http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#import <Foundation/Foundation.h>

#pragma mark - Constants

/** Domain for errors stemming from TwitterLoggingService APIs */
FOUNDATION_EXTERN NSErrorDomain __nonnull const TLSErrorDomain;

/** Pull out an name for the current thread or `nil` if no name was identified */
FOUNDATION_EXTERN NSString * __nullable TLSCurrentThreadName(void);

/**
 These are syslog compatible log levels for use with *TwitterLoggingService*.
 `TLSLog.h` only exposes easy macros for Error, Warning, Information and Debug.

 # Number of levels

 static const NSUInteger TLSLogLevelCount = TLSLogLevelDebug + 1;

 ## Levels to strings

 FOUNDATION_EXTERN NSString *TLSLogLevelToString(TLSLogLevel level);

 */
typedef NS_ENUM(NSInteger, TLSLogLevel)
{
    /** Present for syslog compatability */
    TLSLogLevelEmergency = 0,
    /** Present for syslog compatability */
    TLSLogLevelAlert,
    /** Present for syslog compatability */
    TLSLogLevelCritical,
    /** Use `TLSLogError` (See TLSLog) */
    TLSLogLevelError,
    /** Use `TLSLogWarning` (See TLSLog) */
    TLSLogLevelWarning,
    /** Present for syslog compatability */
    TLSLogLevelNotice,
    /** Use `TLSLogInformation` (See TLSLog) */
    TLSLogLevelInformation,
    /** Use `TLSLogDebug` (See TLSLog) */
    TLSLogLevelDebug
};

//! Number of log levels
static const NSInteger TLSLogLevelCount = TLSLogLevelDebug + 1;

/**
 A set of flags that can be used to identify specific log levels in one mask.
 Used by `TLSOutputStream` conforming objects for filtering log levels.
 */
typedef NS_OPTIONS(NSInteger, TLSLogLevelMask)
{
    /** Emergency */
    TLSLogLevelMaskEmergency   = (1 << TLSLogLevelEmergency),
    /** Alert */
    TLSLogLevelMaskAlert       = (1 << TLSLogLevelAlert),
    /** Critical */
    TLSLogLevelMaskCritical    = (1 << TLSLogLevelCritical),
    /** Error */
    TLSLogLevelMaskError       = (1 << TLSLogLevelError),
    /** Warning */
    TLSLogLevelMaskWarning     = (1 << TLSLogLevelWarning),
    /** Notice */
    TLSLogLevelMaskNotice      = (1 << TLSLogLevelNotice),
    /** Information */
    TLSLogLevelMaskInformation = (1 << TLSLogLevelInformation),
    /** Debug */
    TLSLogLevelMaskDebug       = (1 << TLSLogLevelDebug),

    /** All Levels */
    TLSLogLevelMaskAll         = 0xFF,
    /** No Levels */
    TLSLogLevelMaskNone        = 0,

    /** Emergency and above */
    TLSLogLevelMaskEmergencyAndAbove   = TLSLogLevelMaskEmergency,
    /** Alert and above */
    TLSLogLevelMaskAlertAndAbove       = TLSLogLevelMaskEmergencyAndAbove   | TLSLogLevelMaskAlert,
    /** Critical and above */
    TLSLogLevelMaskCriticalAndAbove    = TLSLogLevelMaskAlertAndAbove       | TLSLogLevelMaskCritical,
    /** Error and above */
    TLSLogLevelMaskErrorAndAbove       = TLSLogLevelMaskCriticalAndAbove    | TLSLogLevelMaskError,
    /** Warning and above */
    TLSLogLevelMaskWarningAndAbove     = TLSLogLevelMaskErrorAndAbove       | TLSLogLevelMaskWarning,
    /** Notice and above */
    TLSLogLevelMaskNoticeAndAbove      = TLSLogLevelMaskWarningAndAbove     | TLSLogLevelMaskNotice,
    /** Information and above */
    TLSLogLevelMaskInformationAndAbove = TLSLogLevelMaskNoticeAndAbove      | TLSLogLevelMaskInformation,
    /** Debug and above (effectively everything except out of bounds values) */
    TLSLogLevelMaskDebugAndAbove       = TLSLogLevelMaskInformationAndAbove | TLSLogLevelMaskDebug
};

/** Advanced options for logging a message */
typedef NS_OPTIONS(NSInteger, TLSLogMessageOptions) {
    /** no options (default behavior) */
    TLSLogMessageOptionsNone = 0,
    /** ignore the `[TLSLoggingService maximumSafeMessageLength]` capping of the message */
    TLSLogMessageOptionsIgnoringMaximumSafeMessageLength = 1 << 0,
};

/**
 Options for how to compose a `TLSLogMessageInfo` into a message string
 Selects which components of the message will be in the composed string.
 All components will format as `@"[TIMESTAMP][THREAD][CHANNEL][LEVEL](__FILE__:__LINE__ __PRETTY_FUNCTION___) : MESSAGE"`
 */
typedef NS_OPTIONS(NSInteger, TLSComposeLogMessageInfoOptions) {
    /**
     No options
     `@" : MESSAGE"`
     */
    TLSComposeLogMessageInfoNoOptions = 0,

    //! TIMESTAMP

    //! Log the `TIMESTAMP` as the time since logging started as *HHH:mm:ss.MMM* (hours, minutes, seconds, milliseconds)
    TLSComposeLogMessageInfoLogTimestampAsTimeSinceLoggingStarted = 1 << 0,
    //! Log the `TIMESTAMP` as the local time as *HHH:mm:ss.MMM* (hours, minutes, seconds, milliseconds)
    TLSComposeLogMessageInfoLogTimestampAsLocalTime = 1 << 1,
    //! Log the `TIMESTAMP` as the UTC time as *HHH:mm:ss.MMM* (hours, minutes, seconds, milliseconds)
    TLSComposeLogMessageInfoLogTimestampAsUTCTime = 1 << 2,

    //! THREAD [THREADNAME] | [THREADID] | [THREADNAME(THREADID)]
    //! Log the `THREAD` identifier
    TLSComposeLogMessageInfoLogThreadId = 1 << 4,
    //! Log the `THREAD` name. @note Take care since thread names can be long and might not be ideal for all logs.
    TLSComposeLogMessageInfoLogThreadName = 1 << 5,

    //! CHANNEL
    //! Log the `CHANNEL`
    TLSComposeLogMessageInfoLogChannel = 1 << 8,

    //! LEVEL
    //! Log the `LEVEL`
    TLSComposeLogMessageInfoLogLevel = 1 << 12,

    //! Callsite Info: (__FILE__:__LINE__ __PRETTY_FUNCTION___)
    //! Log the callsite info always: `(__FILE__:__LINE__ __PRETTY_FUNCTION___)`
    TLSComposeLogMessageInfoLogCallsiteInfoAlways = 1 << 16,
    //! Log the callsite info when message's *level* is `TLSLogLevelWarning` or higher: `(__FILE__:__LINE__ __PRETTY_FUNCTION___)`
    TLSComposeLogMessageInfoLogCallsiteInfoForWarnings = 1 << 17,

    //! Caching
    //! Do not cache the composed log message
    TLSComposeLogMessageInfoDoNotCache = 1 << 31,

    /**
     Default options
     `@"[TIMESTAMP][THREADID][CHANNEL][LEVEL](__FILE__:__LINE__ __PRETTY_FUNCTION___) : MESSAGE"`
     Where `TIMESTAMP` is the time since logging started.
     Where `(__FILE__:__LINE__ __PRETTY_FUNCTION__)` is only present for Warning and above.
     */
    TLSComposeLogMessageInfoDefaultOptions = TLSComposeLogMessageInfoLogTimestampAsTimeSinceLoggingStarted |
                                             TLSComposeLogMessageInfoLogThreadId |
                                             TLSComposeLogMessageInfoLogChannel |
                                             TLSComposeLogMessageInfoLogLevel |
                                             TLSComposeLogMessageInfoLogCallsiteInfoForWarnings,
};

#pragma mark - Declarations

/**
 Encapsulation of log message information.

 All properties of `TLSLogMessageInfo` are readonly and populated at initialization time with the exception of `composeFormattedMessage`.
 */
@interface TLSLogMessageInfo : NSObject

/** The `TLSLogLevel` */
@property (nonatomic, readonly) TLSLogLevel level;
/** The `@__FILE__` of the log message */
@property (nonatomic, nonnull, copy, readonly) NSString *file;
/** The `@__FUNCTION__` of the log message */
@property (nonatomic, nonnull, copy, readonly) NSString *function;
/** The `__LINE__` of the log message */
@property (nonatomic, readonly) NSInteger line;
/** The `NSString*` channel */
@property (nonatomic, nonnull, copy, readonly) NSString *channel;
/** The context object */
@property (nonatomic, nullable, readonly) id contextObject;
/** The log message's timestamp */
@property (nonatomic, nonnull, readonly) NSDate *timestamp;
/** how long the `TLSLoggingService` instance had been alive when this log message was made */
@property (nonatomic, readonly) NSTimeInterval logLifespan;
/** The thread identifier (mach_port_t) that the message was logged from */
@property (nonatomic, readonly) unsigned int threadId;
/** The thread name of the thread that was logged from */
@property (nonatomic, nullable, copy, readonly) NSString *threadName;
/** The log message */
@property (nonatomic, nonnull, copy, readonly) NSString *message;

/**
 Composes a log message in predefined format which is cached for the lifetime of this object.
 @return A log message string using `TLSComposeLogMessageInfoDefaultOptions`
 */
- (nonnull NSString *)composeFormattedMessage;

/**
 Composes a log message in predefined format which is cached for the lifetime of this object.
 @return A log message string using the given `TLSComposeLogMessageInfoOptions` _options_
 */
- (nonnull NSString *)composeFormattedMessageWithOptions:(TLSComposeLogMessageInfoOptions)options;

/**
 Composes a string that combines the _file_, _function_ and _line_ information.
 @return a string in the format `@"(__FILE__:__LINE__ __FUNCTION)"`
 */
- (nonnull NSString *)composeFileFunctionLineString;

/**
 Designated initializer
 */
- (nonnull instancetype)initWithLevel:(TLSLogLevel)level
                                 file:(nonnull NSString *)file
                             function:(nonnull NSString *)function
                                 line:(NSInteger)line
                              channel:(nonnull NSString *)channel
                            timestamp:(nonnull NSDate *)timestamp
                          logLifespan:(NSTimeInterval)logLifespan
                             threadId:(unsigned int)threadId
                           threadName:(nullable NSString *)threadName
                        contextObject:(nullable id)contextObject
                              message:(nonnull NSString *)message NS_DESIGNATED_INITIALIZER;

/**
 `NS_UNAVAILABLE`
 */
- (nonnull instancetype)init NS_UNAVAILABLE;
/**
 `NS_UNAVAILABLE`
 */
+ (nonnull instancetype)new NS_UNAVAILABLE;

@end
